### DISCUSSION TITLE
snes repro and M29W640

### DISCUSSION THREAD

**Discussion by Chomemel** [sanni/cartreader#915] [Need help]
I got a snes repro with a m29w640 instead of a 29lv640. It's a repro of Tales of Phantasia.

I'm not an expert when it comes to datasheet but it seems that the only difference I found between the 2 chips is that the 29w640 has a fast programming mode (but I might be wrong).
I changed the code and put the 29w640 in the 29LV640 group.
When I pluged the cartridge into the OSCR, the chip code detected is C220 and C20B for ExHirom and 83EE and 83A6 for Hirom (I soldered the WE pads and cut the trace between Vcc and WE). Although it should be 20DE or 20DF.
Out of curiosity I tried to detect a 48 pin TSOP 29F160 via the flashrom adapter. The chip is unknown, but good code is detected (018D) by the OSCR.
So whether, flashrom functions work or not, the right code should be detected with a 29w640, if connected correctly to the OSCR, right? 
**Does it mean there is problem with the board (a bridge to solder or remove for exemple)?
Also, how does the detection of the chip code works?**

**Comment by Chomemel** [comment]
Ok, not much progress here, but I noticed something, so I write it down before I forget.

From what I understood from other websites, 29LV640 and 29W640 seem to be programmed the same way via the same adapter.  But I'll know more when I desolder the chip.

I noticed something when I pluged in the OSCR a cartmod of Tales of Phantasia made with a 29LV640. When I go to SNES repro ExHriom, it has exactly the same codes as the repro: C220 and C20B. When I connect the WE of the 29lv640 to the WE of the cartridge, then the 29lv640 is detected by the OSCR (in the snes repro ExHirom menu).

So I suspect that the pcb of the repro prevents from accessing the 29w640. I'll checked again the connexions.

**Comment by Chomemel** [comment]
Today I noticed that 
there is continuity beetween WE and Vdd of the SRAM. This continuity doesn't exist for my Lorom cartridge. What's weird is that if I flip the multimeter's pin, it doesn't bip.
I'll post schemas to illustrate, I just don't has enough time right now.
To be continued...

**Comment by Chomemel** [comment]
I solved my problem by replacing the M29w640 with a 29lv640. So far, erasing, programming, blank checking work great. It boots on RH. I was able do dump the ROM with the OSCR. I still have to test other things (saves, sram replacement...).

As for the 29w640, when I put it in the flashrom adapter it comes with the wrong IDs: 83EE and 83A6 (instead of 20DE and 20DF). I haven't tried CFI when the EEPROM was on the repro, but I guess someone else will do the test as I not really interested in resoldering the 29w640 as the 29LV640 does the job.

**Comment by sanni** [comment]
I added the CFI option to the flashrom programmer menu. 

Tested with the 8bit flash adapter [modified like so](https://github.com/sanni/cartreader/wiki/Flash-Adapter#hw5) and an AMIC A29L160ATV flash chip that is currently not supported by the standard flash mode.

**Reply by Chomemel** [reply]
To be honest, I'm not 100% sure to really understand what CFI  is.  I think I got the purpose of that option, but that's all.

Also, with HW3 revision do I need to modify the flash adapter PCB to make CFI option work?

**Reply by StanMOI** [reply]
CFI = COMMON FLASH MEMORY INTERFACE (CFI) MODE

**Comment by Chomemel** [comment]
I'll use my thread as diary. Or esle I will forgot what I've already done.

After replacing the 29w640 by a 29LV640, here's what I did so far.

**1) Flashing with an ExHirom (Tales of Phantasia FR)**
Flashing, dumping (via flashrom AND SNES), erasing, blankchecking -> ok
playing on RH and emulator (with the dumped ROMs)-> ok
writing save from emulator -> works on RH
duminp save from repro -> works on emulator

**2) Flashing with HiRom (4 Mo)** after cuting the A22 jumper and connecting A22 to the ground.
Wierd and untable results so far, depends on the ROM written. More testing to come

**Comment by Chomemel** [comment]
Here is one weird thing about my repro (now with a 29LV640).

The repro is wired to be a HiRom (A22 connected to the ground). After erasing, here are what I get for the blankcheck:
- if ExHiRom selected in the flash menu: **Flashrom is empty**
- if HiRom selected: **Error: not blank**
I dug deeper and dumped after erasing selecting ExHirom and HiRom. The HiRom file is indeed not empty, but the ExHirom is.

Then, I tried to flash Dragon Quest 6 (4 Mo) selecting HiRom in the flash menu. The writing works but there is an error while verifying. But I dumped the ROM via
- SNES menu, read ROM
- Flashrom Menu (HiRom), read
- Flashrom Menu (ExHirom), read
Dumps via SNES and flashrom ExHirom are identical to the ROM I flashed (the one via the flashrom menu is a 8 Mo file with the ROM written 2 times). The dump via Hirom has some differences (it is also a 8 Mo file). And the repro with DQ6 works on RH.

There seems to be problem for the HiRom menu to treat my repro like a HiRom evan if A22 is not connected.

**Reply by sanni** [reply]
Try todays commit, we found a bug yesterday concerning SRAM interfering with flashing hirom repros.

**Reply by Chomemel** [reply]
Ok, I flash my OSCR with v13.4.





> * if HiRom selected: **Error: not blank**

Now, blankcheck works: **Flashrom is empty** 



>   I dug deeper and dumped after erasing selecting ExHirom and HiRom. The HiRom file is indeed not empty, but the ExHirom is.

The file dumped is now empty both via HiRom and ExHiRom.
 


> Then, I tried to flash Dragon Quest 6 (4 Mo) selecting HiRom in the flash menu. The writing works but there is an error while verifying. 

Gone, now the verifyng process works till the end.



> The dump via Hirom has some differences (it is also a 8 Mo file). And the repro with DQ6 works on RH.
Files are almost identical. 

To sum up, all issues I had, seem to be solved.


Also, I managed to flash the M29W640 via CFI in the FLASHROM PROGRAMMER menu. I haven't tried it, as I have to solder it somewhere. Too bad, I dessoldered it from the repro.

**Reply by StanMOI** [reply]
> Try todays commit, we found a bug yesterday concerning SRAM interfering with flashing hirom repros.

What's new about this BUG ?
I think I have this problem too. Error code when verifying after writing HiROM repros.
[post](https://github.com/sanni/cartreader/discussions/955)
How can I resolve it ?

**Comment by Chomemel** [comment]
I recently replace the 64 ko sram with a 256 ko sram.
I had trouble flashing the eeprom when the jumper A13 and A14 from the sram where bridged to the eeprom/connectors. There where in fact 2 traces that needed to be cut. If the first one was pretty obvious I had trouble figuring there was a second one to cut.
I'll post photos, that might save people some time.
I was able to flash and play the bahamut lagoon translation from Near which is so far the only ExHirom game using 256 ko of Stam.

**Comment by Chomemel** [comment]
I forgot to add that I was able to flash the 29W640 via the [flash adapter](https://github.com/sanni/cartreader/wiki/Flash-Adapter), using the CFI command. Then soldered it to a SNES PCB via an adapter and it works great.

So I was able to use the repro and also to use the M29W640 that was initially on it.
